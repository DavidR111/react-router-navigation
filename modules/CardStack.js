/* @flow */
/* eslint no-duplicate-imports: 0 */

import { Component } from 'react'
import { BackAndroid } from 'react-native'
import { matchPath, withRouter } from 'react-router'
import { StateUtils } from 'react-navigation'
import type { RouterHistory } from 'react-router'
import type { CardRendererProps, NavigationState, Cards, CardProps, CardRoute } from './TypeDefinitions'
import StackUtils from './StackUtils'

type State = {
  navigationState: NavigationState<CardRoute>,
  cards: Cards,
}

type Props = RouterHistory & {
  children: Array<React$Element<CardProps>>,
  render: (props: CardRendererProps) => React$Element<any>,
}

class CardStack extends Component<void, Props, State> {

  props: Props
  state: State

  // Initialyze navigation state with initial history
  constructor(props: Props): void {
    super(props)
    // Build the card stack ($FlowFixMe
    const { children, entries, location } = props
    const cards = children && StackUtils.build(children)
    // Get initial route of navigation state
    if (!entries) throw new Error('No history entries found')
    // Build navigation state
    const navigationState = entries.reduce((state, { pathname }) => {
      const card = cards.find(({ path, exact, strict }) => {
        return matchPath(pathname, path, { exact, strict })
      })
      if (!card || !card.path) return state
      return {
        index: matchPath(location.pathname, card.path, card)
          ? state.routes.length
          : state.index,
        routes: [
          ...state.routes,
          {
            key: StackUtils.createKey(card),
            routeName: card.path,
          },
        ],
      }
    }, { index: -1, routes: [] })
    // Save everything in component state
    this.state = { navigationState, cards }
  }

  // Listen hardware BackAndroid event
  componentDidMount(): void {
    BackAndroid.addEventListener('hardwareBackPress', this.onNavigateBack)
  }

  // Remove all listeners
  componentWillUnmount(): void {
    BackAndroid.removeEventListener('hardwareBackPress', this.onNavigateBack)
  }

  // Listen all history events
  componentWillReceiveProps(nextProps: Props): void {
    // $FlowFixMe
    const { location, action } = nextProps
    const { cards, navigationState } = this.state
    // Get current card
    const currentRoute = navigationState.routes[navigationState.index]
    const currentCard = cards.find(({ key }) => key === currentRoute.routeName)
    // Get next card
    const nextRoute = StackUtils.getRoute(cards, location)
    if (!nextRoute) return
    const nextCard = cards.find(({ key }) => key === nextRoute.key)
    // Local state must be updated ?
    if (
      currentCard && nextCard &&
      StackUtils.shouldUpdate(currentCard, nextCard, this.props, nextProps)
    ) {
      const key = StackUtils.createKey(nextRoute)
      const routeName = nextRoute.key
      switch (action) {
        case 'PUSH': {
          this.setState({
            navigationState: StateUtils.push(
              navigationState,
              { key, routeName },
            ),
          })
          break
        }
        case 'POP': {
          if (this.props.index === undefined || nextProps.index === undefined) return
          const n = this.props.index - nextProps.index
          if (n > 1) {
            this.setState({
              navigationState: StateUtils.reset(
                navigationState,
                navigationState.routes.slice(
                  0,
                  (navigationState.index - n) + 1,
                ),
                navigationState.index - n,
              ),
            })
          } else {
            this.setState({
              navigationState: StateUtils.pop(navigationState),
            })
          }
          break
        }
        case 'REPLACE': {
          this.setState({
            navigationState: StateUtils.replaceAtIndex(
              navigationState,
              navigationState.index,
              { key, routeName },
            ),
          })
          break
        }
        default:
      }
    }
  }

  // Pop to previous scene (n-1)
  onNavigateBack = (): boolean => {
    if (this.state.navigationState.index > 0) {
      this.props.goBack()
      return true
    }
    return false
  }

  // Render view
  render(): React$Element<any> {
    return this.props.render({
      ...this.state,
      onNavigateBack: this.onNavigateBack,
    })
  }

}

export default withRouter(CardStack)
