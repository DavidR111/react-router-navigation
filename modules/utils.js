/* @flow */
/* eslint no-param-reassign: 0 */

import { Children } from 'react'
import { matchPattern } from 'react-router'
import type { NavigationRoute } from 'react-native/Libraries/NavigationExperimental/NavigationTypeDefinition'
import type { Card, MatchCardProps, Tab, MatchTabProps } from './StackTypeDefinitions'

/**
 * Extract current match from all matches
 * and return a route with key prop
 */
export function getCurrentRoute(
  matchs: Array<{ pattern: string, exactly?: boolean } & Object>,
  parent: ?string,
  location: { pathname: string },
): ?NavigationRoute {
  const route = matchs
    .find(({ pattern, exactly }) => {
      return matchPattern(pattern, location, exactly || false, parent)
    })
  if (!route) return null
  return { key: route.pattern }
}


/**
 * Build an item stack by extracting
 * all valid props
 */
export function buildItemStack<Item>(
  children: Array<React$Element<Item>>,
  advancedProps: Array<string>,
  matchComponent: string,
): Array<Item & { key: string }> {
  // Define required props
  const defaultProps = ['component', 'render', 'pattern', 'exactly']
  // Map all children
  return Children.toArray(children).reduce((items, child) => {
    // child === <Match /> || child === <MatchCard />
    const displayName = child.displayName || child.type.displayName || child.type.name
    const isAdvancedMatch = matchComponent === displayName
    if (displayName !== 'Match' && !isAdvancedMatch) return items
    // Build item
    const item = Object.keys(child.props).reduce((props, key) => {
      if (key === 'pattern') {
        props.key = child.props[key]
        props[key] = child.props[key]
      } else if (isAdvancedMatch && (key === 'render' || key === 'component')) {
        props[key] = () => child
      } else if (
        defaultProps.includes(key) ||
        (isAdvancedMatch && advancedProps.includes(key))
      ) {
        props[key] = child.props[key]
      }
      return props
    }, {})
    if (!item.key) return items
    return [...items, item]
  }, [])
}


/**
 * Build a card stack
 */
export function buildCards(children: Array<React$Element<MatchCardProps>>): Array<Card> {
  return buildItemStack(children, ['title', 'hideNavBar'], 'MatchCard')
}


/**
 * Build a tab stack
 */
export function buildTabs(children: Array<React$Element<MatchTabProps>>): Array<Tab> {
  return buildItemStack(children, ['label', 'tabBarStyle'], 'MatchTab')
}
