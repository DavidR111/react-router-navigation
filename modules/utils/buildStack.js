/* @flow */
/* eslint no-param-reassign: 0*/

import { Children, cloneElement } from 'react'

export default function buildStack<Item>(
  children: Array<React$Element<Item>>
): Array<Item & { key: string }> {
  return Children.toArray(children).reduce((stack, child) => {
    const item = Object.keys(child.props).reduce((props, key) => {
      if (key === 'path') {
        props.key = child.props[key]
        props[key] = child.props[key]
      } else if (key === 'render' || key === 'component' || key === 'children') {
        const renderMethod = (ownProps: any) => cloneElement(child, ownProps)
        Object.defineProperty(renderMethod, 'name', { value: key })
        props[key] = renderMethod
      } else {
        props[key] = child.props[key]
      }
      return props
    }, {})
    if (!item.path) return stack
    return [...stack, item]
  }, [])
}
