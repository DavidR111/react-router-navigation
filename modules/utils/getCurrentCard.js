/* @flow */

import type { Route, Cards, Card } from './../TypeDefinitions'
import normalizeRoute from './normalizeRoute'

export default function getCurrentCard(cards: Cards, route: Route): ?Card {
  return cards.find((card) => {
    return normalizeRoute(route).key === card.key
  })
}
