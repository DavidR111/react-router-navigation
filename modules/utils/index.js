/* @flow */

export { default as buildStack } from './buildStack'
export { default as getCurrentCard } from './getCurrentCard'
export { default as getCurrentRoute } from './getCurrentRoute'
export { default as normalizeRoute } from './normalizeRoute'
export { default as shouldStackUpdate } from './shouldStackUpdate'
