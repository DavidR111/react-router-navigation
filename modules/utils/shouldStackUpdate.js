/* @flow */
/* eslint no-duplicate-imports: 0*/

import { matchPath } from 'react-router'
import type { RouterHistory, Location } from 'react-router'
import type { Card } from './../TypeDefinitions'

export default function (
  currentCard: Card,
  nextCard: Card,
  routerHistory: RouterHistory,
  previousLocation: Location,
): boolean {
  const { entries, location, index } = routerHistory
  if (entries === undefined || index === undefined) return false
  const previousEntry = entries[index - 1]
  const currentEntry = entries[index]
  const nextEntry = entries[index + 1]
  const matchCurrentRoute = matchPath(location.pathname, currentCard.path, currentCard)
  const matchNextRoute = matchPath(location.pathname, nextCard.path, nextCard)
  return (
    // URL changes
    (previousLocation.pathname !== location.pathname) &&
    // case 1) basic pathname
    ((currentCard.key !== nextCard.key) ||
    // case 2) Pathname with query params
    // Ex: with same path article/:id,
    //     pathname article/2 !== article/3
    (matchCurrentRoute !== null && matchNextRoute !== null &&
     Object.keys(matchCurrentRoute.params).length !== 0 &&
     Object.keys(matchNextRoute.params).length !== 0 &&
     ((previousEntry && currentEntry.pathname !== previousEntry.pathname) ||
      (nextEntry && currentEntry.pathname !== nextEntry.pathname))
    ))
  )
}
