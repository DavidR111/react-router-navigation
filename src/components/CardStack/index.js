/* @flow */

import React, { Component } from 'react'
import { NavigationExperimental } from 'react-native'
import type { NavigationState, NavigationSceneProps } from './../../types'
import DefaultRenderer from './../DefaultRenderer'
import styles from './styles'

const {
  Transitioner: NavigationTransitioner,
} = NavigationExperimental

type Props = {
  navigationState: NavigationState,
  renderScene: (sceneProps: NavigationSceneProps) => React$Element<any>,
  renderHeader?: (sceneProps: NavigationSceneProps) => React$Element<any> | null,
}

class CardStack extends Component {

  props: Props

  renderScene = (sceneProps: NavigationSceneProps): React$Element<any> => {
    return (
      <DefaultRenderer
        {...this.props}
        {...sceneProps}
        renderScene={this.props.renderScene}
        renderHeader={this.props.renderHeader}
      />
    )
  }

  shouldComponentUpdate(nextProps: Props) {
    return (
      this.props.navigationState.index !== nextProps.navigationState.index ||
      this.props.navigationState.key !== nextProps.navigationState.key
    )
  }

  render() {
    return (
      <NavigationTransitioner
        style={styles.container}
        navigationState={this.props.navigationState}
        render={this.renderScene}
      />
    )
  }

}

export default CardStack
