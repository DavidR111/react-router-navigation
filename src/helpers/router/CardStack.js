/* @flow */

import React, { PropTypes, Component } from 'react'
import { NavigationExperimental, View } from 'react-native'

const {
  Transitioner: NavigationTransitioner,
  CardStack: NavigationCardStack,
  Card: NavigationCard,
} = NavigationExperimental

const {
  CardStackPanResponder,
  CardStackStyleInterpolator,
} = NavigationCard

const {
  Directions,
} = CardStackPanResponder;

export default class Stack extends Component {
  static propTypes = {
    navigationState: PropTypes.object.isRequired,
    style: View.propTypes.style,
    render: PropTypes.func,
  }

  state = {
    inTransition: false,
  }

  renderScenes = (props): React$Element<any> => {
    const { renderOverlay } = this.props
    const overlay = renderOverlay && renderOverlay(props)
    const scenes = props.scenes.map((scene) => (
      this.renderScene({
        ...props,
        scene,
      })
    ))
    return (
      <View style={{ flex: 1 }}>
        <View style={{ flex: 1 }}>
          {scenes}
        </View>
        {overlay}
      </View>
    )
  }

  renderScene = (props): React$Element<any> => {
    const isVertical = this.props.direction === 'vertical'

    const style = isVertical ?
      CardStackStyleInterpolator.forVertical(props) :
      CardStackStyleInterpolator.forHorizontal(props)

    const panHandlersProps = {
      ...props,
      onNavigateBack: this.props.onNavigateBack,
    }
    const panHandlers = isVertical ?
      CardStackPanResponder.forVertical(panHandlersProps) :
      CardStackPanResponder.forHorizontal(panHandlersProps)

    return (
      <NavigationCard
        {...props}
        key={'card_' + props.scene.key}
        panHandlers={panHandlers}
        renderScene={this.props.renderScene}
        style={[style, this.props.cardStyle]}
      />
    )
  }

  render() {
    return (
      <NavigationTransitioner
        navigationState={this.props.navigationState}
        render={(props) => this.renderScenes(props)}
        style={this.props.style}
      />
    )
  }

}
