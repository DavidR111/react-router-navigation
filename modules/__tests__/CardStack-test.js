import { View } from 'react-native'
import React, { PropTypes, Component, createElement } from 'react'
import { Match } from 'react-router'
import { TestRouter, componentFactory, CardView } from './utils'
import CardStack from './../CardStack'
import renderer from 'react-test-renderer'

it('<CardStack /> renders correctly', () => {
  const component = renderer.create(
    <TestRouter>
      <CardStack render={CardView}>
        <Match exactly pattern="/" component={componentFactory('Index')} />
        <Match pattern="/hello" component={componentFactory('Hello')} />
      </CardStack>
    </TestRouter>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<CardStack /> renders correctly with initialIndex and initialEntries prop ', () => {
  const component = renderer.create(
    <TestRouter
      initialIndex={1}
      initialEntries={['/', '/hello', '/goodbye']}
    >
      <CardStack render={CardView}>
        <Match exactly pattern="/" component={componentFactory('Index')} />
        <Match pattern="/hello" component={componentFactory('Hello')} />
        <Match pattern="/goodbye" component={componentFactory('Goodbye')} />
      </CardStack>
    </TestRouter>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<CardStack /> re-renders correctly when "push" action is called', () => {
  const component = renderer.create(
    <TestRouter>
      <CardStack render={CardView}>
        <Match exactly pattern="/" component={componentFactory('Index')} />
        <Match pattern="/hello" component={componentFactory('Hello')} />
      </CardStack>
    </TestRouter>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  tree.props.onPress((history) => history.push('/hello'))
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<CardStack /> re-renders correctly when "pop" action is called', () => {
  const component = renderer.create(
    <TestRouter
      initialIndex={1}
      initialEntries={['/', '/hello']}
    >
      <CardStack render={CardView}>
        <Match exactly pattern="/" component={componentFactory('Index')} />
        <Match pattern="/hello" component={componentFactory('Hello')} />
      </CardStack>
    </TestRouter>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  tree.props.onPress((history) => history.goBack())
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<CardStack /> re-renders correctly when "replace" action is called', () => {
  const component = renderer.create(
    <TestRouter
      initialIndex={1}
      initialEntries={['/', '/hello']}
    >
      <CardStack render={CardView}>
        <Match exactly pattern="/" component={componentFactory('Index')} />
        <Match pattern="/hello" component={componentFactory('Hello')} />
        <Match pattern="/goodbye" component={componentFactory('Goodbye')} />
      </CardStack>
    </TestRouter>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  tree.props.onPress((history) => history.replace('/goodbye'))
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<CardStack /> re-renders correctly when onNavigateBack() method is called', () => {
  const component = renderer.create(
    <TestRouter
      initialIndex={1}
      initialEntries={['/', '/hello']}
    >
      <CardStack render={({ navigationState, onNavigateBack, cards }) => {
        return (
          <View onPress={onNavigateBack}>
            {CardView({ navigationState, onNavigateBack, cards })}
          </View>
        )
      }}>
        <Match exactly pattern="/" component={componentFactory('Index')} />
        <Match pattern="/hello" component={componentFactory('Hello')} />
      </CardStack>
    </TestRouter>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  tree.children[0].children[0].props.onPress()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
