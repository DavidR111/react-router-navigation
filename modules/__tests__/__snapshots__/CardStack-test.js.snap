exports[`test <CardStack /> re-renders correctly when "pop" action is called 1`] = `
<View
  onPress={[Function]}>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Hello
    </Text>
  </View>
</View>
`;

exports[`test <CardStack /> re-renders correctly when "pop" action is called 2`] = `
<View
  onPress={[Function]}>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Index
    </Text>
  </View>
</View>
`;

exports[`test <CardStack /> re-renders correctly when "push" action is called 1`] = `
<View
  onPress={[Function]}>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Index
    </Text>
  </View>
</View>
`;

exports[`test <CardStack /> re-renders correctly when "push" action is called 2`] = `
<View
  onPress={[Function]}>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Hello
    </Text>
  </View>
</View>
`;

exports[`test <CardStack /> re-renders correctly when "replace" action is called 1`] = `
<View
  onPress={[Function]}>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Hello
    </Text>
  </View>
</View>
`;

exports[`test <CardStack /> re-renders correctly when "replace" action is called 2`] = `
<View
  onPress={[Function]}>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Goodbye
    </Text>
  </View>
</View>
`;

exports[`test <CardStack /> re-renders correctly when onNavigateBack() method is called 1`] = `
<View
  onPress={[Function]}>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }>
    <View
      onPress={[Function]}>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail">
        Hello
      </Text>
    </View>
  </View>
</View>
`;

exports[`test <CardStack /> re-renders correctly when onNavigateBack() method is called 2`] = `
<View
  onPress={[Function]}>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }>
    <View
      onPress={[Function]}>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail">
        Index
      </Text>
    </View>
  </View>
</View>
`;

exports[`test <CardStack /> renders correctly 1`] = `
<View
  onPress={[Function]}>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Index
    </Text>
  </View>
</View>
`;

exports[`test <CardStack /> renders correctly with initialIndex and initialEntries prop  1`] = `
<View
  onPress={[Function]}>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Hello
    </Text>
  </View>
</View>
`;
